File: swift-bridge/Sources/EventKitBridge/Object.swift
import EventKit
import Foundation

@_cdecl("ek_object_from_event_json")
public func ek_object_from_event_json(
    _ store: UnsafeMutableRawPointer?,
    _ eventJSON: UnsafePointer<CChar>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutableRawPointer? {
    guard let store else {
        ekrSetMessageError(outError, message: "missing EKEventStore")
        return nil
    }

    do {
        let payload = try ekrDecodeJSON(eventJSON, as: EKREventPayload.self)
        let event = try ekrPrepareEvent(
            store: ekrBorrow(store, as: EKEventStore.self),
            payload: payload,
            requireCalendar: false
        )
        return ekrRetain(event as EKObject)
    } catch {
        ekrSetError(outError, error)
        return nil
    }
}

@_cdecl("ek_object_from_reminder_json")
public func ek_object_from_reminder_json(
    _ store: UnsafeMutableRawPointer?,
    _ reminderJSON: UnsafePointer<CChar>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutableRawPointer? {
    guard let store else {
        ekrSetMessageError(outError, message: "missing EKEventStore")
        return nil
    }

    do {
        let payload = try ekrDecodeJSON(reminderJSON, as: EKRReminderPayload.self)
        let reminder = try ekrPrepareReminder(
            store: ekrBorrow(store, as: EKEventStore.self),
            payload: payload,
            requireCalendar: false
        )
        return ekrRetain(reminder as EKObject)
    } catch {
        ekrSetError(outError, error)
        return nil
    }
}

@_cdecl("ek_object_from_calendar_draft_json")
public func ek_object_from_calendar_draft_json(
    _ store: UnsafeMutableRawPointer?,
    _ calendarJSON: UnsafePointer<CChar>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutableRawPointer? {
    guard let store else {
        ekrSetMessageError(outError, message: "missing EKEventStore")
        return nil
    }

    do {
        let payload = try ekrDecodeJSON(calendarJSON, as: EKRCalendarDraftPayload.self)
        let calendar = try ekrPrepareCalendar(
            store: ekrBorrow(store, as: EKEventStore.self),
            payload: payload,
            requireSource: false
        )
        return ekrRetain(calendar as EKObject)
    } catch {
        ekrSetError(outError, error)
        return nil
    }
}

@_cdecl("ek_object_release")
public func ek_object_release(_ object: UnsafeMutableRawPointer?) {
    guard let object else { return }
    ekrRelease(object)
}

@_cdecl("ek_object_has_changes")
public func ek_object_has_changes(_ object: UnsafeMutableRawPointer?) -> Bool {
    guard let object else { return false }
    return ekrBorrow(object, as: EKObject.self).hasChanges
}

@_cdecl("ek_object_is_new")
public func ek_object_is_new(_ object: UnsafeMutableRawPointer?) -> Bool {
    guard let object else { return false }
    return ekrBorrow(object, as: EKObject.self).isNew
}

@_cdecl("ek_object_reset")
public func ek_object_reset(_ object: UnsafeMutableRawPointer?) {
    guard let object else { return }
    ekrBorrow(object, as: EKObject.self).reset()
}

@_cdecl("ek_object_rollback")
public func ek_object_rollback(_ object: UnsafeMutableRawPointer?) {
    guard let object else { return }
    ekrBorrow(object, as: EKObject.self).rollback()
}

@_cdecl("ek_object_refresh")
public func ek_object_refresh(_ object: UnsafeMutableRawPointer?) -> Bool {
    guard let object else { return false }
    return ekrBorrow(object, as: EKObject.self).refresh()
}

File: src/object.rs
//! EventKit live-object wrappers.

use core::ffi::{c_char, c_void};
use std::fmt;
use std::ptr::{self, NonNull};

use crate::calendar::EKCalendarDraft;
use crate::error::EventKitError;
use crate::event::EKEvent;
use crate::event_store::EKEventStore;
use crate::ffi;
use crate::private::json_cstring;
use crate::reminder::EKReminder;

/// Wraps a live EventKit `EKObject` instance.
pub struct EKObject {
    raw: NonNull<c_void>,
}

impl fmt::Debug for EKObject {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("EKObject").finish_non_exhaustive()
    }
}

impl EKObject {
    fn from_raw(
        raw: *mut c_void,
        error: *mut c_char,
        context: &str,
    ) -> Result<Self, EventKitError> {
        NonNull::new(raw).map(|raw| Self { raw }).ok_or_else(|| {
            if error.is_null() {
                EventKitError::OperationFailed(format!("failed to create {context}"))
            } else {
                unsafe { EventKitError::from_error_ptr(error, &format!("{context} failed")) }
            }
        })
    }

    /// Wraps a live EventKit event as `EKObject`.
    pub fn from_event(store: &EKEventStore, event: &EKEvent) -> Result<Self, EventKitError> {
        let payload = json_cstring(event, "EKEvent")?;
        let mut error = ptr::null_mut();
        let raw = unsafe {
            ffi::object::ek_object_from_event_json(store.as_raw_ptr(), payload.as_ptr(), &mut error)
        };
        Self::from_raw(raw, error, "EKObject from EKEvent")
    }

    /// Wraps a live EventKit reminder as `EKObject`.
    pub fn from_reminder(
        store: &EKEventStore,
        reminder: &EKReminder,
    ) -> Result<Self, EventKitError> {
        let payload = json_cstring(reminder, "EKReminder")?;
        let mut error = ptr::null_mut();
        let raw = unsafe {
            ffi::object::ek_object_from_reminder_json(
                store.as_raw_ptr(),
                payload.as_ptr(),
                &mut error,
            )
        };
        Self::from_raw(raw, error, "EKObject from EKReminder")
    }

    /// Wraps a live EventKit calendar draft as `EKObject`.
    pub fn from_calendar_draft(
        store: &EKEventStore,
        calendar: &EKCalendarDraft,
    ) -> Result<Self, EventKitError> {
        let payload = json_cstring(calendar, "EKCalendarDraft")?;
        let mut error = ptr::null_mut();
        let raw = unsafe {
            ffi::object::ek_object_from_calendar_draft_json(
                store.as_raw_ptr(),
                payload.as_ptr(),
                &mut error,
            )
        };
        Self::from_raw(raw, error, "EKObject from EKCalendarDraft")
    }

    /// Returns whether the wrapped EventKit object has unsaved changes.
    pub fn has_changes(&self) -> bool {
        unsafe { ffi::object::ek_object_has_changes(self.raw.as_ptr()) }
    }

    /// Returns whether the wrapped EventKit object is new to EventKit.
    pub fn is_new(&self) -> bool {
        unsafe { ffi::object::ek_object_is_new(self.raw.as_ptr()) }
    }

    /// Resets the wrapped EventKit object to its last committed state.
    pub fn reset(&self) {
        unsafe { ffi::object::ek_object_reset(self.raw.as_ptr()) };
    }

    /// Rolls back pending EventKit changes on the wrapped object.
    pub fn rollback(&self) {
        unsafe { ffi::object::ek_object_rollback(self.raw.as_ptr()) };
    }

    /// Refreshes the wrapped EventKit object from the store.
    pub fn refresh(&self) -> bool {
        unsafe { ffi::object::ek_object_refresh(self.raw.as_ptr()) }
    }
}

impl Drop for EKObject {
    fn drop(&mut self) {
        unsafe { ffi::object::ek_object_release(self.raw.as_ptr()) };
    }
}
